src/errors/validation_exception.rs

```diff
@@ -94,6 +115,91 @@ impl ValidationError {
     pub fn use_default_error() -> PyErr {
         py_schema_error_type!("Uncaught UseDefault error, please check your usage of `default` validators.")
     }
+
+    fn maybe_add_cause(self_: PyRef<'_, Self>, py: Python) -> Option<PyErr> {
+        let mut user_py_errs = vec![];
+        for line_error in &self_.line_errors {
+            if let ErrorType::AssertionError {
+                error: Some(err),
+                context: _,
+            }
+            | ErrorType::ValueError {
+                error: Some(err),
+                context: _,
+            } = &line_error.error_type
+            {
+                let note: PyObject = if let Location::Empty = &line_error.location {
+                    "Pydantic: cause of loc: root".into_py(py)
+                } else {
+                    format!(
+                        "Pydantic: cause of loc: {}",
+                        // Location formats with a newline at the end, hence the trim()
+                        line_error.location.to_string().trim()
+                    )
+                    .into_py(py)
+                };
+
+                // Notes only support 3.11 upwards:
+                #[cfg(Py_3_11)]
+                {
+                    // Add the location context as a note, no direct c api for this,
+                    // fine performance wise, add_note() goes directly to C: "(PyCFunction)BaseException_add_note":
+                    // https://github.com/python/cpython/blob/main/Objects/exceptions.c
+                    if err.call_method1(py, "add_note", (format!("\n{note}"),)).is_ok() {
+                        user_py_errs.push(err.clone_ref(py));
+                    }
+                }
+
+                // Pre 3.11 notes support, use a UserWarning exception instead:
+                #[cfg(not(Py_3_11))]
+                {
+                    use pyo3::exceptions::PyUserWarning;
+
+                    let wrapped = PyUserWarning::new_err((note,));
+                    wrapped.set_cause(py, Some(PyErr::from_value(err.as_ref(py))));
+                    user_py_errs.push(wrapped);
+                }
```

Comment:
I'm curious for the choice to wrap in a `UserWarning`?

Reply:
Pre 3.11 we can't use a note to specify the error location, so we need to do it in an error, I figured considering its not an error, a "warning" made the most sense, happy to change if you've got something better!

Reply:
Given that pre-3.11 users won't be reading `__notes__`, think it's maybe better to just not emit notes for them? That way we don't emit a different exception type on different versions. (`ValueError` vs `UserWarning`)